### PR TITLE
:sparkles: Add support for specifying TLS config including custom CA certificates

### PR DESCRIPTION
**Pull request by jimmidyson**
**What this PR does / why we need it**:
Currently it is very difficult to use CAAPH with charts from an OCI registry served by a registry with a custom CA. This PR fixes that by allowing a CA certificate secret to be specified on a HelmChartProxy which will ultimately be used by the helm client to install/upgrade the cgart.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by jimmidyson** [comment]
The e2e tests are failing due to calico images - fix in #239 which needs to be merged before this one.

**Comment by jimmidyson** [comment]
Rebased to pull in calico fix from #239.

File: controllers/helmreleaseproxy/helmreleaseproxy_controller.go [hunk 1]
```diff
@@ -224,21 +224,39 @@ func (r *HelmReleaseProxyReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+		defer func() {
+			if err := os.Remove(credentialsPath); err != nil {
+				log.Error(err, "failed to remove credentials file in path", "credentialsPath", credentialsPath)
+			}
+		}()
+	}
+
+	caFilePath, err := r.getCAFile(ctx, helmReleaseProxy)
+	if err != nil {
+		wrappedErr := errors.Wrapf(err, "failed to get CA file for cluster")
```
Comment (mboersma) [inline controllers/helmreleaseproxy/helmreleaseproxy_controller.go:237]:
```suggestion
		wrappedErr := errors.Wrapf(err, "failed to get CA certificate file for cluster")
```
Just so we're using the same terminology "CA certificate" in both these error message.

File: controllers/helmreleaseproxy/helmreleaseproxy_controller.go [hunk 2]
```diff
@@ -377,6 +395,31 @@ func (r *HelmReleaseProxyReconciler) getCredentials(ctx context.Context, helmRel
 	return credentialsPath, nil
 }
 
+// getCAFile fetches the CA certificate from a Secret and writes them to a temporary file it returns the path to the temporary file.
```
Comment (mboersma) [inline controllers/helmreleaseproxy/helmreleaseproxy_controller.go:398]:
```suggestion
// getCAFile fetches the CA certificate from a Secret and writes it to a temporary file, returning the path to the temporary file.
```

**Review by mboersma — COMMENTED** [review]
lgtm, thanks again @jimmidyson! I just had two minor comments on doc strings.

/assign @Jont828

**Comment by faiq** [comment]
Does this require any RBAC permission changes to read a secret across different namespaces?

File: controllers/helmreleaseproxy/helmreleaseproxy_controller.go (hunk 1 above)
Comment (jimmidyson) [reply controllers/helmreleaseproxy/helmreleaseproxy_controller.go:237]: Thanks!

File: controllers/helmreleaseproxy/helmreleaseproxy_controller.go (hunk 2 above)
Comment (jimmidyson) [reply controllers/helmreleaseproxy/helmreleaseproxy_controller.go:398]: Thanks!

**Review by mboersma — COMMENTED** [review]
/lgtm

I'd like to get another maintainer's eyes on this as well, ping @jackfrancis @Jont828.

File: api/v1alpha1/helmchartproxy_types.go [hunk 3]
```diff
@@ -199,6 +202,16 @@ type Credentials struct {
 	Key string `json:"key"`
 }
 
+type TLSConfig struct {
```
Comment (Jont828) [inline api/v1alpha1/helmchartproxy_types.go:205]:
```suggestion
// TLSConfig defines a TLS configuration.
type TLSConfig struct {
```

File: api/v1alpha1/helmchartproxy_types.go [hunk 4]
```diff
@@ -199,6 +202,16 @@ type Credentials struct {
...
+	// Secret is a reference to a Secret containing the TLS CA certificate at the key ca.crt.
+	// +optional
+	CASecretRef *corev1.SecretReference `json:"caSecret,omitempty"`
+
+	// InsecureSkipTLSverify controls whether the Helm client should verify the server's certificate.
+	// +optional
+	InsecureSkipTLSverify bool `json:"insecureSkipTLSVerify,omitempty"`
```
Comment (Jont828) [inline api/v1alpha1/helmchartproxy_types.go:212]:
```suggestion
	InsecureSkipTLSVerify bool `json:"insecureSkipTLSVerify,omitempty"`
```

File: api/v1alpha1/helmchartproxy_types.go [hunk 5]
```diff
@@ -199,6 +202,16 @@ type Credentials struct {
 	Key string `json:"key"`
 }
 
+type TLSConfig struct {
+	// Secret is a reference to a Secret containing the TLS CA certificate at the key ca.crt.
+	// +optional
+	CASecretRef *corev1.SecretReference `json:"caSecret,omitempty"`
+
+	// InsecureSkipTLSverify controls whether the Helm client should verify the server's certificate.
```
Comment (Jont828) [inline api/v1alpha1/helmchartproxy_types.go:210]:
```suggestion
	// InsecureSkipTLSVerify controls whether the Helm client should verify the server's certificate.
```

File: api/v1alpha1/helmchartproxy_types.go [hunk 6]
```diff
@@ -72,6 +72,9 @@ type HelmChartProxySpec struct {
 	// Credentials is a reference to an object containing the OCI credentials. If it is not specified, no credentials will be used.
 	// +optional
 	Credentials *Credentials `json:"credentials,omitempty"`
+
+	// TLSConfig contains the TLS configuration for the HelmChartProxy.
```
Comment (Jont828) [inline api/v1alpha1/helmchartproxy_types.go:76]:
```suggestion
	// TLSConfig contains the TLS configuration for a HelmChartProxy.
	// +optional
```

Should this be optional if it's omitted when empty?

**Review by Jont828 — COMMENTED** [review]
Thanks for the patience, been busy with several projects recently. Overall seems fine, just some nits regarding the API types.

File: api/v1alpha1/helmchartproxy_types.go (hunk 6 above)
Comment (jimmidyson) [reply api/v1alpha1/helmchartproxy_types.go:76]: Yes, using both `+optional` and `omitempty` is idiomatic - see https://github.com/kubernetes/community/blob/1114fe76ba5de7d38da786e900fc9b715d76441f/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: api/v1alpha1/helmchartproxy_types.go (hunk 3 above)
Comment (jimmidyson) [reply api/v1alpha1/helmchartproxy_types.go:206]: Done.

File: api/v1alpha1/helmchartproxy_types.go (hunk 4 above)
Comment (jimmidyson) [reply api/v1alpha1/helmchartproxy_types.go]: Done.

**Comment by jimmidyson** [comment]
Thanks @Jont828! Applied review feedback and pushed 🙏

File: api/v1alpha1/helmchartproxy_types.go (hunk 6 above)
Comment (Jont828) [reply api/v1alpha1/helmchartproxy_types.go]: Alright in that case can you set TLSConfig to be `+optional`?

Comment (Jont828) [reply api/v1alpha1/helmchartproxy_types.go]: I think we should be good to merge after that!

File: api/v1alpha1/helmchartproxy_types.go [hunk 7]
```diff
@@ -72,6 +72,9 @@ type HelmChartProxySpec struct {
 	// Credentials is a reference to an object containing the OCI credentials. If it is not specified, no credentials will be used.
 	// +optional
 	Credentials *Credentials `json:"credentials,omitempty"`
+
+	// TLSConfig contains the TLS configuration for a HelmChartProxy.
+	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
```
Comment (Jont828) [inline api/v1alpha1/helmchartproxy_types.go:77]:
```suggestion
        // +optional
	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
```

File: api/v1alpha1/helmchartproxy_types.go (hunk 6 above)
Comment (jimmidyson) [reply api/v1alpha1/helmchartproxy_types.go]: Oops sorry I thought I had!

**Comment by faiq** [comment]
@Jont828 can we get a release for this functionality :pray:

**Review by mboersma — COMMENTED** [review]
/lgtm

**Comment by Jont828** [comment]
Can you squash? LGTM otherwise. And sure, I can cut a new release after we get the current PRs merged.

**Comment by jimmidyson** [comment]
Thanks @Jont828! Squashed and ready to go 🚀

**Pull request merged**